client/test/e2e.test.js

```diff
@@ -0,0 +1,112 @@
+import test from 'tape';
+import puppeteer from 'puppeteer';
+
+function sleep(ms) {
+    return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+const width = 1920;
+const height = 1080;
+
+function browser(width, height) {
+    return puppeteer.launch({
+    headless: true,
+    args: [`--window-size=${width},${height}`],
+    // slowMo: 20 // For testing
+    });
+}
+
+test('e2e-client-test-valid-actions', async (t) => {
+    // Arrange: Set up the page
+    const b = await browser(width, height);
+    const p = await b.newPage();
+    await p.setViewport({ width, height });
+    await p.goto("http://127.0.0.1:8080/");
+    
+    // Act: Test creating a workspace
+    await p.waitForSelector("#add-workspace");
+    await p.click("#add-workspace")
+    await p.waitForSelector("#workspace-name");
+    await p.focus('#workspace-name')
+    await p.keyboard.type('puppeteer')
+    await p.click("#create-workspace")
+
+    // Assert: Check that the new workspace exists with no tables
+    await p.waitForSelector(".v-list-item");
+    await p.waitForSelector(".v-list-item__title");
+    const workspace_name = await p.evaluate(() => document.querySelector('.v-list-item__title').innerText);
+    t.equal(workspace_name, "puppeteer", "The new workspace is created and called the right thing.")
+
+    // Assert: Check that there are no tables or graphs yet
+    await p.waitForSelector(".ws-detail-empty-list");
+    let tables = await p.evaluate(() => document.querySelectorAll('.ws-detail-empty-list')[0].innerText.split("info ")[1]);
```

Comment:
Since this is repeated *almost* verbatim below, it might make sense to create a pair of functions to actually describe what this is doing.

Furthermore, this evaluation will likely cause a crash if for some reason there were already tables in the workspace. To improve this, you can first check the `length` property of the result of `querySelectorAll()` before pulling out the content.

Reply:
Moving this into a function for cleaner code, but under what circumstances would there be tables that exist in the workspace? The testing code uses the test environment from the client and the API and wipes it all out before it loads. Do you mean in case we add more tests first?

Reply:
"In case we add more tests first" would be a definite way this could happen, but my general attitude about this is more defensive in nature: when testing a system, you should make as few assumptions as possible. So asserting that the table list is empty is a nice way to do that, and it prepares the formal precondition for the test that follows.

I agree it's an unlikely scenario, and one that we could fix if it ever did come up. But since it's such a small fix now, it's one of those things we could just consider doing and forgetting about.

Reply:
Gotcha. This is the section I'm having issues with

Reply:
Fixed in 1e4c955